src/torchcodec/decoders/_core/DeviceInterface.h

```diff
@@ -23,25 +24,41 @@ namespace facebook::torchcodec {
 //   deviceFunction(device, ...);
 // }
 
-// Initialize the hardware device that is specified in `device`. Some builds
-// support CUDA and others only support CPU.
-void initializeContextOnCuda(
-    const torch::Device& device,
-    AVCodecContext* codecContext);
-
-void convertAVFrameToFrameOutputOnCuda(
-    const torch::Device& device,
-    const VideoDecoder::VideoStreamOptions& videoStreamOptions,
-    UniqueAVFrame& avFrame,
-    VideoDecoder::FrameOutput& frameOutput,
-    std::optional<torch::Tensor> preAllocatedOutputTensor = std::nullopt);
-
-void releaseContextOnCuda(
-    const torch::Device& device,
-    AVCodecContext* codecContext);
-
-std::optional<const AVCodec*> findCudaCodec(
-    const torch::Device& device,
-    const AVCodecID& codecId);
+struct DeviceInterface {
+  DeviceInterface(const std::string& device) : device_(device) {}
+
+  virtual ~DeviceInterface(){};
+
+  torch::Device& device() {
+    return device_;
+  };
+
+  virtual std::optional<const AVCodec*> findCodec(const AVCodecID& codecId) = 0;
+
+  // Initialize the hardware device that is specified in `device`. Some builds
+  // support CUDA and others only support CPU.
+  virtual void initializeContext(AVCodecContext* codecContext) = 0;
+
+  virtual void convertAVFrameToFrameOutput(
+      const VideoDecoder::VideoStreamOptions& videoStreamOptions,
+      UniqueAVFrame& avFrame,
+      VideoDecoder::FrameOutput& frameOutput,
+      std::optional<torch::Tensor> preAllocatedOutputTensor = std::nullopt) = 0;
+
+  virtual void releaseContext(AVCodecContext* codecContext) = 0;
```

Comment:
Do we need this to be a named member function? That is, isn't it appropriate for this to just happen in the destructor?

Reply:
Note that as currently defined `DeviceInterface` does not store `AVCodecContext`. It merely initializes it on a call to `initializeContext()`. i.e. `AVCodecContext* codecContext` argument in this method is both input and output one.

Note also that not storying of `AVCodecContext` is convenient if we potentially want same `DeviceInterface` capable to handle multiple contexts at once. Do we?

Reply:
On a second thought, I removed `releaseContext`. Actually I overlooked that you have `StreamInfo` where you store stream specific stuff. So, couple changes laid quite nicely after I released that. Including no need for `releaseContext` and `std::unique_ptr` for `DeviceInterface`.